File: api/src/main/java/com/fabio/flashcards_app/domain/services/AuthService.java
package com.fabio.flashcards_app.domain.services;

import com.fabio.flashcards_app.data.dto.auth.AuthResponseDTO;
import com.fabio.flashcards_app.data.dto.auth.LoginDTO;
import com.fabio.flashcards_app.data.dto.auth.RegisterDTO;
import com.fabio.flashcards_app.domain.models.User;
import com.fabio.flashcards_app.domain.repositories.UserRepository;
import com.fabio.flashcards_app.security.JwtService;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class AuthService {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private JwtService jwtService;

    @Autowired
    private PasswordEncoder passwordEncoder;

    public AuthResponseDTO register(RegisterDTO dto) {
        if(userRepository.findByEmail(dto.email()).isPresent()) {
            throw new RuntimeException("Email already in use");
        }

        User user = new User();
        user.setName(dto.name());
        user.setEmail(dto.email());
        // Hash BCrypt before save
        user.setPassword(passwordEncoder.encode(dto.password()));
        user.setRole("USER");

        userRepository.save(user);

        String token = jwtService.generateToken(user);
        return new AuthResponseDTO(
                token,
                user.getName(),
                user.getEmail(),
                user.getRole()
        );
    }

    public AuthResponseDTO login(LoginDTO dto) {
        User user = userRepository.findByEmail(dto.email())
                .orElseThrow(() -> new RuntimeException("Email not found"));

        // auth password
        if (!passwordEncoder.matches(dto.password(), user.getPassword())) {
            throw new RuntimeException("Wrong password");
        }

        String token = jwtService.generateToken(user);

        return new AuthResponseDTO(
                token,
                user.getName(),
                user.getEmail(),
                user.getRole()
        );
    }
}

File: web/src/services/auth.service.ts
import api from './api'

export const authService = {
  login: async (email: string, password: string) => {
    const { data } = await api.post('/auth/login', { email, password })

    const user = { name: data.name, email: data.email, role: data.role }
    
    localStorage.setItem('token', data.token)
    localStorage.setItem('user', JSON.stringify(user))

    return { token: data.token, user }
  },
  register: async (name: string, email: string, password: string) => {
    const { data } = await api.post('/auth/register', { name, email, password })
    const user = { name: data.name, email: data.email, role: data.role }
    localStorage.setItem('token', data.token)
    localStorage.setItem('user', JSON.stringify(user))
    return { token: data.token, user }
  },
  logout: () => {
    localStorage.removeItem('token')
    localStorage.removeItem('user')
  }
}
